### DISCUSSION TITLE
how do i Create Keys And Certificate on android device using aws-iot-device-sdk-java-v2, we used to create that aws-iot-device-sdk-java-v1

### DISCUSSION THREAD

**Discussion by nilesh-borkar** [aws/aws-iot-device-sdk-java-v2#605] [Q&A] [answered]
Hi 

Does anyone know how to create keys and Certificate on device, we were able to create the key and certificate like this in version 1 IOT SDK - 

```
CreateKeysAndCertificateRequest createKeysAndCertificateRequest =
                                    new CreateKeysAndCertificateRequest();
                            createKeysAndCertificateRequest.setSetAsActive(true);
                            final CreateKeysAndCertificateResult createKeysAndCertificateResult;
                            createKeysAndCertificateResult =
                                    mIotAndroidClient.createKeysAndCertificate(createKeysAndCertificateRequest);
                                     AWSIotKeystoreHelper.saveCertificateAndPrivateKey(mCertificateId,
                                    createKeysAndCertificateResult.getCertificatePem(),
                                    createKeysAndCertificateResult.getKeyPair().getPrivateKey(),
                                    mKeystorePath, mKeystoreName, mKeystorePassword);
                                    AttachPrincipalPolicyRequest policyAttachRequest =
                                    new AttachPrincipalPolicyRequest();
                            policyAttachRequest.setPolicyName(mAWSIoTPolicyName);
                            policyAttachRequest.setPrincipal(createKeysAndCertificateResult.getCertificateArn());
                            mIotAndroidClient.attachPrincipalPolicy(policyAttachRequest);
```
 
but now from version 2 IOT SDK few classes are not available and we are not able to create this information on device level,
any help would be much appreaciated 

Thanks in advance

**Comment by jmklix** [comment, marked as answer]
Sorry, but you can't create keys and certificates with this sdk. You would have to do that manually or with the [aws cli](https://docs.aws.amazon.com/iot/latest/developerguide/device-certs-create.html). Please let me know if you have any questions about how to do this

**Reply by KamleshAtTray** [reply]
The FleetProvisioning sample mentions generating keys and a certificate using a provisioning claim. 

https://github.com/aws/aws-iot-device-sdk-java-v2/tree/main/samples/FleetProvisioning#creating-a-certificate-key-set-from-a-provisioning-claim

What connection method should be used if we cannot perform the provisioning steps in advance, such as generating certificates/keys using the AWS CLI and provisioning the claim on IoT devices (which, in our case, are Android devices)?

We want our application to handle provisioning programmatically and connect to AWS IoT autonomously.

**Reply by bretambrose** [reply]
If you're not able to deploy any kind of provisioning material (initial cert/key or AWS credentials) to the device, your best bet is probably a cognito-based work flow (https://aws.amazon.com/cognito/).  The user would need to create an account through your webapp and use those credentials to login.  The results/token of that login can be supplied to a Cognito credentials provider (https://github.com/awslabs/aws-crt-java/blob/main/src/main/java/software/amazon/awssdk/crt/auth/credentials/CognitoCredentialsProvider.java) allowing you to establish an MQTT connection over websockets.

**Reply by KamleshAtTray** [reply]
Thank you @bretambrose We're going ahead with Cognito method.

**Comment by bretambrose** [comment]
I would also add:

I don't know where you're getting the code snippet you've included, but the v1 Device SDK for Java does not include any of the types or APIs in the snippet.  The v1 device SDK has never supported this.  If I had to guess, I'd say you're confusing the AWS v1 SDK for Java and the IoT Device SDK for Java.
